src/kv_client/request.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::{
+    kv_client::{HasError, KvClient, RpcFnType, Store},
+    pd::PdClient,
+    Result,
+};
+
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use futures::stream::BoxStream;
+use grpcio::CallOption;
+use kvproto::kvrpcpb;
+use std::sync::Arc;
+
+pub trait KvRequest: Sync + Send + 'static + Sized + Clone {
+    type Result;
+    type RpcRequest;
+    type RpcResponse: HasError + Clone + Send + 'static;
+    /// A single `KvRequest` can be divided into a number of RPC requests because the keys span
+    /// several regions or a single RPC request is too large. Most of the fields in these requests
+    /// share the same content while `Payload` is the part which differs among the requests.
+    type Payload;
+    const REQUEST_NAME: &'static str;
+    const RPC_FN: RpcFnType<Self::RpcRequest, Self::RpcResponse>;
+
+    fn execute(
+        mut self,
+        pd_client: Arc<impl PdClient>,
+    ) -> BoxFuture<'static, Result<Self::Result>> {
+        let stores = self.store_stream(pd_client);
+        Self::reduce(
+            stores
+                .and_then(move |(key, store)| {
+                    let request = self.into_request(key, &store);
+                    self.mock_dispatch(&request, store.call_options())
+                        .unwrap_or_else(|| store.dispatch::<Self>(&request, store.call_options()))
+                })
+                .map_ok(move |r| Self::map_result(r))
+                .boxed(),
+        )
+    }
+
+    fn store_stream<PdC: PdClient>(
+        &mut self,
+        pd_client: Arc<PdC>,
+    ) -> BoxStream<'static, Result<(Self::Payload, Store<PdC::KvClient>)>>;
+
+    fn into_request<KvC: KvClient>(
```

Comment:
This should not be `into_` if it takes `self` by reference. Maybe `make_rpc_request`?

Reply:
Sounds good.